src/js/CheckboxTree.js

```diff
@@ -81,129 +80,153 @@ class CheckboxTree extends React.Component {
-
-        this.state = {
-            id: props.id || `rct-${nanoid(7)}`,
-            model,
-            prevProps: props,
-        };
-
-        this.onCheck = this.onCheck.bind(this);
-        this.onExpand = this.onExpand.bind(this);
-        this.onNodeClick = this.onNodeClick.bind(this);
-        this.onExpandAll = this.onExpandAll.bind(this);
-        this.onCollapseAll = this.onCollapseAll.bind(this);
-    }
-
-    // eslint-disable-next-line react/sort-comp
-    static getDerivedStateFromProps(newProps, prevState) {
-        const { model, prevProps } = prevState;
-        const { disabled, id, nodes } = newProps;
-        let newState = { ...prevState, prevProps: newProps };
+        let stateChanged = false;
 
-        // Apply new properties to model
-        model.setProps(newProps);
+        if (id && id !== state.id) {
+            newState.id = id;
+            stateChanged = true;
+        }
 
-        // Since flattening nodes is an expensive task, only update when there is a node change
-        if (!isEqual(prevProps.nodes, nodes) || prevProps.disabled !== disabled) {
-            model.flattenNodes(nodes);
+        if (icons !== CheckboxTree.defaultProps.icons && icons !== state.prevIconsProp) {
+            let iconsChanged = false;
+            const newIcons = { ...state.icons };
+            const keys = Object.keys(icons);
+            keys.forEach((key) => {
+                if (!state.icons[key] || icons[key] !== state.icons[key]) {
+                    iconsChanged = true;
+                    newIcons[key] = icons[key];
+                }
+            });
+            if (iconsChanged) {
+                newState.icons = newIcons;
+                newState.prevIconsProp = icons;
+                stateChanged = true;
+            }
         }
 
-        if (id !== null) {
-            newState = { ...newState, id };
+        if (stateChanged) {
+            return newState;
         }
-        model.deserializeLists({
-            checked: newProps.checked,
-            expanded: newProps.expanded,
-        });
 
-        return newState;
+        return null;
     }
 
-    onCheck(nodeInfo) {
-        const { checkModel, noCascade, onCheck } = this.props;
-        const model = this.state.model.clone();
-        const node = model.getNode(nodeInfo.value);
-
-        model.toggleChecked(nodeInfo, nodeInfo.checked, checkModel, noCascade);
-        onCheck(model.serializeList('checked'), { ...node, ...nodeInfo });
+    state = {
+        id: `rct-${nanoid(7)}`,
+        icons: { ...CheckboxTree.defaultProps.icons },
+        prevIconsProp: {},
     }
 
-    onExpand(nodeInfo) {
-        const { onExpand } = this.props;
-        const model = this.state.model.clone();
-        const node = model.getNode(nodeInfo.value);
+    // This instance variable is used to store the parent of each node (keyed on node.value)
+    // for use when updating nodes during onCheck(), onExpand() and updateParentNodes().
+    parents = {}
```

Comment:
This is an instance variable because it is loaded in `renderTreeNodes` as the nodes are processed. Since `renderTreeNodes` is part of `render` and you do not change `state` in `render` I made `parents` and instance variable. It is an instance variable because it is needed during `onCheck` after rendering.

This is part of getting rid of the flat nodes which required multiple walks of the tree. We are storing as little as possible in state and only walking the tree once. Since we have to walk the tree in render every time, we only do it there and let `parents` be a calculated value. Even with the large data tree, this happens quickly. The slow down is the changing of the DOM and I've done my best to make sure nothing changes if not needed. The `render` code of `CheckboxTree` seems to run quickly even for the large data example.
 
Since the tree structure may be changed outside of `CheckboxTree` having this calculated in the render pass is better that trying to store it in state. Figuring out if it has changed is expensive.

That is my reasoning.  Let me know if you see a flaw in it.